protocol/0064-VALP-validator_performance_based_rewards.md

```diff
@@ -36,6 +52,50 @@ be modified and set to neutral defaults (i.e., 0)
 
 Then `validator_performance = max(0.05, min((p'/expected, 1))`
 
+Goal 2: Detect unforwarded Ethereum events and lack or read access and punish the validator that does not forward them
+Detection: Events forwarded by some validators are not forwarded by others.
+
+#### Ethereum Heartbeat
+
+For the Ethereum Heartbeat, we use the network parameter `ethereum_heartbeat_period`. This parameter should be either 0 or a value bigger than the number of validators; the recommended initial value is 128, which would create a hearbeat per validator about every 20 minutes (i.e., about 120 heartbeats per validator per epoch). Legal values are all integers larger or equal to 0.
+
+For every Ethereum block, if the hash of this block mod `ethereum_heartbeat_period` equals the identity of the a validator (taken mod ethereum_heartbeat_period)+1, then this validator has to forward this as an Ethereum event. This event is confirmed by other validators just like any other Ethereum event, but then ignored. If that block also contains a valid Vega event that requires an action, this is forwarded independently by the normal event forwarding mechanisms. The heartbeat also does contain the Ethereum hash of the corresponding block.
+If the parameter is set to 0, the heartbeats are effectively turned off.
+
+If a validator forwards a heartbeat, all other validators have to validate its correctness by checking if the hashes work out.
+
+#### Ethereum Read Access Heartbeat
```

Comment:
Just to clarify, what is the purpose of having both `Ethereum Heartbeat` and `Ethereum Read Access Heartbeat`?

Reply:
Ethereum heartbeat measures if you follow the Ethereum chain to spot events happening there that need forwarding (e.g., deposits on the bridge). Read access means you have access to the Ethereum state, so you can get oracle readings and such.

Reply:
So do we expect the listen for events for the first mechanism and pull for the other?

Reply:
Yes